### PR TITLE
Optimize AAE replication.

### PR DESCRIPTION
**Pull request by cmeiklejohn**
Implement a prototype optimization for active anti-entropy replication which front-loads the first two levels of bucket retrieval through one TCP request, instead of 1025 roundtrips.  Futher optimizations include increasing the number of levels which are prefetched.

This is to generally vet the changeset, and explore possible ways of mitigating the protocol fallback and backwards compatability.

### CODE REVIEW

**Comment by cmeiklejohn** [comment]
cc: @Vagabond

File: src/riak_repl_aae_fullsync.hrl [hunk 1]
```diff
@@ -3,6 +3,7 @@
 -define(MSG_UPDATE_TREE, 3).
 -define(MSG_GET_AAE_BUCKET, 4).
 -define(MSG_GET_AAE_SEGMENT, 5).
+-define(MSG_GET_AAE_BUCKETS, 6).
```
Comment (Vagabond) [inline src/riak_repl_aae_fullsync.hrl]: Aren't you overloading message code 6 here?

Comment (cmeiklejohn) [inline src/riak_repl_aae_fullsync.hrl]: Yeah, I totally did.

File: src/riak_repl_aae_source.erl [hunk 2]
```diff
@@ -280,7 +285,18 @@ key_exchange(start_key_exchange, State=#state{cluster=Cluster,
                              ok
                      end;
                 (get_bucket, {L, B}) ->
-                     send_synchronous_msg(?MSG_GET_AAE_BUCKET, {L,B,IndexN}, State);
+                     lager:warning("Bucket request for ~p ~p",
```
Comment (Vagabond) [inline src/riak_repl_aae_source.erl]: This is presumably debugging?

File: src/riak_repl_aae_source.erl [hunk 3]
```diff
@@ -280,7 +285,18 @@ key_exchange(start_key_exchange, State=#state{cluster=Cluster,
...
+                                   [L, B]),
+                     case orddict:find({L, B}, Buckets) of
```
Comment (Vagabond) [inline src/riak_repl_aae_source.erl]: We fetch all the buckets, don't we? There should never be a cache miss here.

**Comment by Vagabond** [comment]
+1, repl_aae_fullsync passes.

**Comment by cmeiklejohn** [comment]
Closing; not merging.

**Pull request closed without merging**